### PR TITLE
refactor: metrics

### PR DESCRIPTION
**Pull request by Arqu**
A bit bigger one:
- moved most of the metrics definitions from individual crates into metrics - much easier to navigate + can share them more easily through other crates
- cleaned out a lot of the cruft
- added resolver specific metrics regarding cache hit/miss, propagate that into metadata
- pushed metrics down into the loader to capture size & time metrics
- general cleanup stuff

### CODE REVIEW

File: iroh-resolver/src/resolver.rs [hunk 1]
```diff
@@ -225,22 +229,33 @@ pub enum UnixfsType {
 }
 
 pub enum OutPrettyReader<T: ContentLoader> {
-    DagPb(usize, Bytes),
+    DagPb(usize, Bytes, iroh_metrics::gateway::Metrics, Instant),
```
Comment (dignifiedquire) [inline iroh-resolver/src/resolver.rs:232]: these are getting quite annoying, lets put this into its own type, sth like

```rust
DagPb(BytesReader),

struct BytesReader {
  pos: usize,
  bytes: Bytes,
  metrics: Metrics,
  start: Instant,
}
```

File: iroh-resolver/src/resolver.rs [hunk 2]
```diff
@@ -225,22 +229,33 @@ pub enum UnixfsType {
...
+    Raw(usize, Bytes, iroh_metrics::gateway::Metrics, Instant),
 }
 
 impl Out {
-    pub fn pretty<T: ContentLoader>(self, loader: T) -> OutPrettyReader<T> {
+    pub fn pretty<T: ContentLoader>(
+        self,
+        loader: T,
+        metrics: iroh_metrics::gateway::Metrics,
+        start_time: Instant,
```
Comment (dignifiedquire) [inline iroh-resolver/src/resolver.rs:244]: This is really messy having to pass the time manually around :/ can't we store that in the `Metrics` struct maybe?

File: iroh-resolver/src/resolver.rs [hunk 3]
```diff
@@ -252,54 +267,86 @@ impl<T: ContentLoader + Unpin + 'static> AsyncRead for OutPrettyReader<T> {
...
+#[derive(Debug)]
+pub struct LoadedCid {
+    pub data: Bytes,
+    pub source: Source,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Source {
+    Bitswap,
+    Store(String),
```
[the remark this answers is not in the document]
Comment (Arqu) [reply iroh-resolver/src/resolver.rs:307]: :( how?
dropping in a &str requires a boatload of lifetimes to be passed around.

[the remark this answers is not in the document]
Comment (dignifiedquire) [reply iroh-resolver/src/resolver.rs:307]: that's why I said `static`

```rust
struct Foo {
  my_static_str: &'static str,
}
```

will compile fine

**Review by dignifiedquire — DISMISSED** [review]

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**